README.md

```diff
@@ -71,109 +113,88 @@ This script writes 5 files:
+would re-download unnecessary documents again.
 
-Note: unfortunately, loading the old URLs takes a very long time... also, the
-script runs in a single process as Python's shared memory performance is
+Note: unfortunately, loading the previously collected URLs takes a very long time... 
+also, the script runs in a single process as Python's shared memory performance is
 abysmal.
 
 ### Download pages
 
-Pages in the (filtered, deduplicated) index can be downloaded by the command
-```
-download_pages.py -o 2019/cc_downloaded -e warc.gz -i '2019/cc_index_dedup/*.gz'
-```
+Common crawl offers to ways to download their corpus: via http and via Amazon's
+S3 buckets. The first version is getting less and less support and since 2023 it
+has been throttled down. So we switched over to use the S3 version.
 
-This step takes a while, so it make sense to [distribute the work among
-a cluster of machines](#tech). However, it is a bit more involved than
-distributing other scripts, as this one has been inherited from the old
-repository. The differences are:
+This requires a valid AWS account to be configured using ENV variables with the
+default names: AWS_ACCESS_KEY, AWS_SECRET_ACCES_KEY, etc.
 
-1. The script requires a `glob` (possibly wildcard) expression that expands to
-   a list of input files, and not an input directory. So the place of the host
-   name in the value to the `-i` argument must be specified manually (with
-   `{}`, see below)
-1. The number of processes must be set to 0. This eliminates the `-P` argument
-   from the Python command line. We need to do this because first, `-P` means
-   something else in `download_pages.py`, and second, the script uses threads
-   anyway.
+Thanks to Amazon's support for the Common Crawl project, downloading is free.
 
+Pages in the (filtered, deduplicated) index can be downloaded by the command
 ```
-ansible-playbook -i hosts distribute_files.yml -e
-    '{"input_dir": "/mnt/data/lang/Hungarian/cc_corpus/2019/cc_index_dedup",
-      "output_dir": "/mnt/data/lang/Hungarian/cc_corpus/2019"}'
-
-ansible-playbook -i hosts python.yml -e
-    '{"python_script": "download_pages.py",
-      "log_file": "2019_download.log",
-      "working_dir": "/mnt/data/lang/Hungarian/cc_corpus/",
-      "arguments": "-o $output_dir -e warc.gz -i $input_glob",
-      "per_host_args": {"input_glob": "\"2019/cc_index_dedup_{}/*.gz\"",
-                        "output_dir": "2019/cc_downloaded/"}, "processes": 0}'
+download_pages.py -i 03_index_dedup/2019-04 -o 04_downloaded/2019-04
+--index_output_dir 04a_index_sorted/2019-04 --error_file 04b_download_errors
 ```
 
+It is highly recommended to set the number of processes (-P) as high as
+possible in your environment. The reason for this is that S3 does not allow 
+the download of multiple slices from a single file during a single request.
+This means that the number of download requests we make will be comparable
+to the total number of documents we download.
```

Comment:
True, but what about downthrottling? What is the situation with that?

Reply:
S3 documentation actively encourages using many threads for situations like these, I have not encountered downthrottling on their side.

Reply:
OK, not downthrottling then, but aren't we getting SlowDown requests for the index right now?

Reply:
We even got "SlowDown" right at the very first request...